src/charm.py

```diff
@@ -0,0 +1,1017 @@
+        logger.info("Retrieved database details")
+
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Hydra container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
+            return
+
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        if not self._get_secrets():
+            self.unit.status = WaitingStatus("Waiting for secret creation")
+            event.defer()
+            return
+
+        if not self._migration_is_needed():
+            self._handle_status_update_config(event)
+            return
+
+        if not self.unit.is_leader():
+            logger.info("Unit does not have leadership")
+            self.unit.status = WaitingStatus("Unit waiting for leadership to run the migration")
+            event.defer()
+            return
+
+        if not self._run_sql_migration():
+            self.unit.status = BlockedStatus("Database migration job failed")
+            logger.error("Automigration job failed, please use the run-migration action")
+            return
+
+        self._set_peer_data(self._migration_peer_data_key, self._hydra_cli.get_version())
+        self._handle_status_update_config(event)
+
+    def _on_database_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event Handler for database changed event."""
+        self._handle_status_update_config(event)
+
+    def _on_run_migration(self, event: ActionEvent) -> None:
+        """Runs the migration as an action response."""
+        if not self._container.can_connect():
+            event.fail("Service is not ready. Please re-run the action when the charm is active")
+            return
+
+        timeout = float(event.params.get("timeout", 120))
+        event.log("Migrating database.")
+        try:
+            self._hydra_cli.run_migration(timeout=timeout, dsn=self._dsn)
+        except Error as e:
+            err_msg = e.stderr if isinstance(e, ExecError) else e
+            event.fail(f"Database migration action failed: {err_msg}")
+            return
+        event.log("Successfully migrated the database.")
+
+        if not self._peers:
+            event.fail("Peer relation not ready. Failed to store migration version")
+            return
+        self._set_peer_data(self._migration_peer_data_key, self._hydra_cli.get_version())
+        event.log("Updated migration version in peer data.")
+
+        if self.unit.status == WaitingStatus("Waiting for migration to run"):
+            self._handle_status_update_config(event)
+
+    def _on_database_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Event Handler for database relation departed event."""
+        self.unit.status = BlockedStatus("Missing required relation with postgresql")
```

Comment:
As mentioned before, I think maybe `relation-broken` might be a better choice here as if the database scale down an unit, it will trigger the `relation-departed` event.